File: packages/sdk-ts/src/cache.ts
// ─────────────────────────────────────────────
//  PromptOps SDK — Cache
//  In-memory LRU-style cache for fetched prompts.
//  Reduces latency for hot paths in production.
// ─────────────────────────────────────────────

import type { RawPrompt } from './types'

interface CacheEntry {
  prompt: RawPrompt
  expiresAt: number
}

export class PromptCache {
  private store = new Map<string, CacheEntry>()
  /** Cache TTL in milliseconds. Default: 5 minutes */
  private ttl: number

  constructor(ttlMs = 5 * 60 * 1000) {
    this.ttl = ttlMs
  }

  key(slug: string, env: string, version?: string): string {
    return version ? `${slug}::${version}` : `${slug}::env:${env}`
  }

  get(slug: string, env: string, version?: string): RawPrompt | null {
    const entry = this.store.get(this.key(slug, env, version))
    if (!entry) return null
    if (Date.now() > entry.expiresAt) {
      this.store.delete(this.key(slug, env, version))
      return null
    }
    return entry.prompt
  }

  set(slug: string, env: string, prompt: RawPrompt, version?: string): void {
    this.store.set(this.key(slug, env, version), {
      prompt,
      expiresAt: Date.now() + this.ttl,
    })
  }

  invalidate(slug: string): void {
    for (const key of this.store.keys()) {
      if (key.startsWith(`${slug}::`)) {
        this.store.delete(key)
      }
    }
  }

  clear(): void {
    this.store.clear()
  }

  size(): number {
    return this.store.size
  }
}

File: packages/sdk-py/promptops/cache.py
# ─────────────────────────────────────────────
#  PromptOps Python SDK — Cache
# ─────────────────────────────────────────────

from __future__ import annotations
import time
from typing import Optional
from .types import RawPrompt


class PromptCache:
    """In-memory cache for fetched prompts with TTL expiry."""

    def __init__(self, ttl_seconds: float = 300.0):
        self._store: dict[str, tuple[RawPrompt, float]] = {}
        self._ttl = ttl_seconds

    def _key(self, slug: str, env: str, version: Optional[str] = None) -> str:
        return f"{slug}::v:{version}" if version else f"{slug}::env:{env}"

    def get(self, slug: str, env: str, version: Optional[str] = None) -> Optional[RawPrompt]:
        key = self._key(slug, env, version)
        entry = self._store.get(key)
        if not entry:
            return None
        prompt, expires_at = entry
        if time.monotonic() > expires_at:
            del self._store[key]
            return None
        return prompt

    def set(self, slug: str, env: str, prompt: RawPrompt, version: Optional[str] = None) -> None:
        key = self._key(slug, env, version)
        self._store[key] = (prompt, time.monotonic() + self._ttl)

    def invalidate(self, slug: str) -> None:
        keys_to_delete = [k for k in self._store if k.startswith(f"{slug}::")]
        for key in keys_to_delete:
            del self._store[key]

    def clear(self) -> None:
        self._store.clear()

    def size(self) -> int:
        return len(self._store)